x/sync/network_server.go

```diff
@@ -136,32 +141,100 @@ func (s *NetworkServer) HandleChangeProofRequest(
 	}
 
 	// override limit if it is greater than maxKeyValuesLimit
-	limit := req.Limit
-	if limit > maxKeyValuesLimit {
-		limit = maxKeyValuesLimit
+	keyLimit := req.KeyLimit
+	if keyLimit > maxKeyValuesLimit {
+		keyLimit = maxKeyValuesLimit
 	}
+	bytesLimit := int(req.BytesLimit)
+	if bytesLimit > constants.DefaultMaxMessageSize {
+		bytesLimit = constants.DefaultMaxMessageSize
+	}
+
+	// attempt to get a proof within the bytes limit
+	for keyLimit > 0 {
+		changeProof, err := s.db.GetChangeProof(ctx, req.StartingRoot, req.EndingRoot, req.Start, req.End, int(keyLimit))
+		if err != nil {
+			// handle expected errors so clients cannot cause servers to spam warning logs.
+			if errors.Is(err, merkledb.ErrRootIDNotPresent) || errors.Is(err, merkledb.ErrStartRootNotFound) {
+				s.log.Debug(
+					"dropping invalid change proof request",
+					zap.Stringer("nodeID", nodeID),
+					zap.Uint32("requestID", requestID),
+					zap.Stringer("req", req),
+					zap.Error(err),
+				)
+				return nil // dropping request
+			}
+			return err
+		}
+
+		proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
+		if err != nil {
+			return err
+		}
+		if len(proofBytes) < bytesLimit {
+			return s.appSender.SendAppResponse(ctx, nodeID, requestID, proofBytes)
+		}
+		// the proof size was too large, try to shrink it
 
-	changeProof, err := s.db.GetChangeProof(ctx, req.StartingRoot, req.EndingRoot, req.Start, req.End, int(limit))
-	if err != nil {
-		// handle expected errors so clients cannot cause servers to spam warning logs.
-		if errors.Is(err, merkledb.ErrRootIDNotPresent) || errors.Is(err, merkledb.ErrStartRootNotFound) {
-			s.log.Debug(
-				"dropping invalid change proof request",
+		// ensure that the new limit is always smaller
+		keyLimit = uint16((len(changeProof.KeyValues) + len(changeProof.DeletedKeys)) / 2)
+
+		// estimate the bytes of the start and end proof to ensure that everything will fit into the bytesLimit
+		bytesEstimate := getBytesEstimateOfProofNodes(changeProof.StartProof)
+
+		// just the start proof is too large, so a proof is impossible
+		if bytesEstimate > int(req.BytesLimit) {
+			// errors are fatal, so log for the moment
+			s.log.Warn(
+				"cannot generate a proof within bytes limit",
 				zap.Stringer("nodeID", nodeID),
 				zap.Uint32("requestID", requestID),
 				zap.Stringer("req", req),
-				zap.Error(err),
+				zap.Error(ErrMinProofSizeIsTooLarge),
 			)
-			return nil // dropping request
+			return nil
 		}
-		return err
-	}
 
-	proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
-	if err != nil {
-		return err
+		bytesEstimate += getBytesEstimateOfProofNodes(changeProof.EndProof)
```

Comment:
Should we add some buffer to this to account for the fact that when we adjust the key limit, the new end proof may be larger than the current end proof?

Reply:
What amount of buffer do you think would be sufficient?  Theoretically the new end proof could be arbitrarily larger than the current end proof.